### DISCUSSION TITLE
CodeSmell - Reforzar conceptos

### DISCUSSION THREAD

**Discussion by wormeno** [dds-jv/foro#134] [Consultas de objetos] [answered]
Hola,
   Quiero reforzar un poco la idea de CodeSmell, con un código donde veo muchas cosas mezcladas: es del trabajo (saco nombres propios usados):

    class Clase1 extends ClasePadre implements Interface1,Interface2{
       interface NuevaInterface extends ClasePadre2 , ClasePadre3 {
    	   void metodo1(Documento doc, String cod);
    	   void metodo1(String codSMP);
	   void metodo1(String area, double d, boolean b);
      }
}
   Observo los siguientes "CodeSmell"

- Missing polymorphism. No se entiende de qué tipo es un objeto: pues Clase 1 es subclase e implementa otras. Además si tengo una nueva clase que extiende de ClasePadre: no queda claro el comportamiento que debe tener. 
- Refused Bequest. No usar éste tipo de herencias múltiples, tal vez se deba usar delegación. 

   Y algo que no termino de entender es la interface dentro de la clase:
         interface NuevaInterface extends ClasePadre2 , ClasePadre3 {
    	      void metodo1(Documento doc, String cod);
    	      void metodo1(String codSMP);
	      void metodo1(String area, double d, boolean b);
         }
    

- Definir una interface que es subclase de otras dos clases
- Definir tres métodos con el mismo nombre donde varía el tipo y cantidad de argumentos. Googleando encontré que esto es llamando "Sobrecarga de métodos". Esto funciona pero esto no es expresivo ni declarativo.

   Claramente todo éste código funciona, pero hace demasiado ruido.
   Bueno quería compartir el código y el análisis que hago del mismo en función de los conceptos aprendidos de la materia.
   Por favor corregirme si en algún punto estoy aplicando mal los conceptos, y si alguien puede explicar el uso de una interface dentro de una clase lo agradecería (de hecho, esta definición plantuml no lo reconoce!!)
   Gracias

**Comment by gastonprieto** [comment]
Esta larga, pero me parece bien, mañana te respondo en detalle, pero antes me adelanto con dos cosas:

* no todos los smells tienen nombre (después reviso los que me comentaste por las dudas)
* esa interface que tenes ahí se usa en algun lugar?

**Comment by gastonprieto** [comment]
> Missing polymorphism.

En realidad el code smell es *type test*, lo de *missing polymorphism* es como la causa, o sea hiciste una type test porque falta polimorfismo probablemente.

Mas allá de eso esto lo evidencias en el uso del objeto, no lo podes ver en la declaración de la clase.

> No se entiende de qué tipo es un objeto: pues Clase 1 es subclase e implementa otras. Además si tengo una nueva clase que extiende de ClasePadre: no queda claro el comportamiento que debe tener.

Esto la verdad no entiendo muy bien a lo que apuntas, o sea,  los mensajes que entienda son los que implementa de la interfaz + los que hereda + los que uno defina. O sea, eso es lo único que te interesa desde afuera, que mensajes expone (su interfaz)

**Comment by gastonprieto** [comment]
> Refused Bequest. No usar éste tipo de herencias múltiples, tal vez se deba usar delegación.

No entiendo a que te referís con que usa herencia múltiple, en Java la herencia es simple.

**Comment by gastonprieto** [comment]
> Y algo que no termino de entender es la interface dentro de la clase:
```Java
interface NuevaInterface extends ClasePadre2 , ClasePadre3 {
void metodo1(Documento doc, String cod);
void metodo1(String codSMP);
void metodo1(String area, double d, boolean b);
}
```

No tengo idea porque es interna, habría que ver como se usa.

> Definir tres métodos con el mismo nombre donde varía el tipo y cantidad de argumentos. Googleando encontré que esto es llamando "Sobrecarga de métodos". Esto funciona pero esto no es expresivo ni declarativo.

Si, se llama sobrecarga de métodos en general no soy fan de esa soluciones, ya que suelen a llevar a complejizar el uso, y muchas veces denota una falta de alguna abstracción.

Yo no diria que tiene que ver con la declaratividad, que en general esta muy asociada a la idea de ocultar un algoritmo repetible cross dominio.

**Comment by gastonprieto** [comment]
> Definir una interface que es subclase de otras dos clases

Ojo, no es una subclase, esa es la forma de decir que una interfaz cumple el contrato de otras interfaces, o sea, sirve para ampliar el contrato en uno mas grande. Aunque en general se recomienda tener contratos mas chicos. podría llegar a servir en algunos contextos.

**Comment by roli93** [comment]
@wormeno Te sirvió la respuesta?

**Comment by wormeno** [comment, marked as answer]
Gracias Gaston,
Es un código de un sistema en el trabajo que me tocó revisar. Actualmente está funcionando. Yo le saqué los nombre originales, pero la estructura es tal cual como la plateé.
En un principio me costó entenderlo. 
Me quedó claro que no existe herencia múltiple. Y que la interface "interna" la usan para extender el contrato.
En general me parece un código bastante feo y poco declarativo.
Pero bueno en el mundo laboral uno se va encontrando con diversas formas de programar.
Gracias por la explicación.
